File: 0200-number-of-islands/0200-number-of-islands.cpp
class Solution {
public:
    void findIsland(vector<vector<char>>& grid, vector<vector<bool>>& seen, int i, int j){
        if(grid[i][j] == '0' || seen[i][j] == true) return;
        seen[i][j] = true;
        if(i+1 < grid.size()) findIsland(grid, seen, i+1, j);
        if(j+1 < grid[0].size()) findIsland(grid,seen, i, j+1);
        if(i-1 >= 0) findIsland(grid, seen, i-1, j);
        if(j-1 >= 0) findIsland(grid, seen, i, j-1);
    }

    int numIslands(vector<vector<char>>& grid) {
        vector<vector<bool>> seen(grid.size(), vector<bool>(grid[0].size(), false));
        int res = 0;
        for(int i = 0; i < grid.size(); i++){
            for(int j = 0; j < grid[0].size(); j++){
                if(seen[i][j] || grid[i][j] == '0'){
                    continue;
                }else{
                    findIsland(grid, seen, i, j);
                    res++;
                }
            }
        }
        return res;
    }
};

File: 0200-number-of-islands/0200-number-of-islands.py
from collections import deque
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        m = len(grid)
        n = len(grid[0])
        count = 0
        for i in range(m):
            for j in range(n):
                if grid[i][j] == "1":
                    queue = deque([(i,j)])
                    grid[i][j] = "0"
                    while queue:
                        r,c = queue.popleft()
                        if r - 1 >= 0 and grid[r-1][c] == "1":
                            grid[r-1][c] = "0"
                            queue.append([r-1, c])
                        if r + 1 < m and grid[r+1][c] == "1":
                            grid[r+1][c] = "0"
                            queue.append([r+1, c])
                        if c -1 >= 0 and grid[r][c-1] == "1":
                            grid[r][c-1] = "0"
                            queue.append([r, c-1])
                        if c + 1 < n and grid[r][c+1] == "1":
                            grid[r][c+1] = "0"
                            queue.append([r, c+1])
                    count += 1
        return count